File: Server~/src/tools/createPrimitiveTool.ts
import * as z from 'zod';
import { Logger } from '../utils/logger.js';
import { McpUnity } from '../unity/mcpUnity.js';
import { McpServer } from '@modelcontextprotocol/sdk/server/mcp.js';
import { McpUnityError, ErrorType } from '../utils/errors.js';
import { CallToolResult } from '@modelcontextprotocol/sdk/types.js';

const toolName = 'create_primitive';
const toolDescription = 'Creates a primitive GameObject (Cube, Sphere, Capsule, Cylinder, Plane, Quad) in the current scene';
const paramsSchema = z.object({
  primitiveType: z.enum(['Cube', 'Sphere', 'Capsule', 'Cylinder', 'Plane', 'Quad']).describe('The type of primitive to create'),
  name: z.string().optional().describe('Optional name for the GameObject (defaults to the primitive type name)'),
  parentPath: z.string().optional().describe('Optional path of a parent GameObject in the hierarchy'),
  parentId: z.number().optional().describe('Optional instance ID of a parent GameObject'),
  position: z.object({
    x: z.number().optional(),
    y: z.number().optional(),
    z: z.number().optional()
  }).optional().describe('Optional local position'),
  rotation: z.object({
    x: z.number().optional(),
    y: z.number().optional(),
    z: z.number().optional()
  }).optional().describe('Optional local rotation in euler angles'),
  scale: z.object({
    x: z.number().optional(),
    y: z.number().optional(),
    z: z.number().optional()
  }).optional().describe('Optional local scale (defaults to 1,1,1)')
});

export function registerCreatePrimitiveTool(server: McpServer, mcpUnity: McpUnity, logger: Logger) {
  logger.info(`Registering tool: ${toolName}`);

  server.tool(
    toolName,
    toolDescription,
    paramsSchema.shape,
    async (params: any) => {
      try {
        logger.info(`Executing tool: ${toolName}`, params);
        const result = await toolHandler(mcpUnity, params);
        logger.info(`Tool execution successful: ${toolName}`);
        return result;
      } catch (error) {
        logger.error(`Tool execution failed: ${toolName}`, error);
        throw error;
      }
    }
  );
}

async function toolHandler(mcpUnity: McpUnity, params: any): Promise<CallToolResult> {
  const response = await mcpUnity.sendRequest({
    method: toolName,
    params: {
      primitiveType: params.primitiveType,
      name: params.name,
      parentPath: params.parentPath,
      parentId: params.parentId,
      position: params.position,
      rotation: params.rotation,
      scale: params.scale
    }
  });

  if (!response.success) {
    throw new McpUnityError(
      ErrorType.TOOL_EXECUTION,
      response.message || 'Failed to create primitive'
    );
  }

  return {
    content: [{
      type: response.type,
      text: response.message
    }]
  };
}

File: Editor/Tools/CreatePrimitiveTool.cs
using System;
using McpUnity.Unity;
using McpUnity.Utils;
using UnityEngine;
using UnityEditor;
using Newtonsoft.Json.Linq;

namespace McpUnity.Tools
{
    public class CreatePrimitiveTool : McpToolBase
    {
        public CreatePrimitiveTool()
        {
            Name = "create_primitive";
            Description = "Creates a primitive GameObject (Cube, Sphere, Capsule, Cylinder, Plane, Quad) in the current scene";
        }

        public override JObject Execute(JObject parameters)
        {
            string primitiveTypeStr = parameters["primitiveType"]?.ToObject<string>();
            string name = parameters["name"]?.ToObject<string>();
            string parentPath = parameters["parentPath"]?.ToObject<string>();
            int? parentId = parameters["parentId"]?.ToObject<int?>();
            JObject position = parameters["position"] as JObject;
            JObject rotation = parameters["rotation"] as JObject;
            JObject scale = parameters["scale"] as JObject;

            if (string.IsNullOrEmpty(primitiveTypeStr))
            {
                return McpUnitySocketHandler.CreateErrorResponse(
                    "Required parameter 'primitiveType' not provided",
                    "validation_error"
                );
            }

            if (!Enum.TryParse<PrimitiveType>(primitiveTypeStr, true, out PrimitiveType primitiveType))
            {
                return McpUnitySocketHandler.CreateErrorResponse(
                    $"Invalid primitive type '{primitiveTypeStr}'. Valid types: Cube, Sphere, Capsule, Cylinder, Plane, Quad",
                    "validation_error"
                );
            }

            GameObject primitive = GameObject.CreatePrimitive(primitiveType);
            Undo.RegisterCreatedObjectUndo(primitive, $"Create {primitiveType}");

            if (!string.IsNullOrEmpty(name))
            {
                primitive.name = name;
            }

            if (!string.IsNullOrEmpty(parentPath) || parentId.HasValue)
            {
                GameObject parent = null;
                if (parentId.HasValue)
                {
                    parent = EditorUtility.InstanceIDToObject(parentId.Value) as GameObject;
                }
                else
                {
                    parent = GameObject.Find(parentPath);
                }

                if (parent != null)
                {
                    Undo.SetTransformParent(primitive.transform, parent.transform, $"Parent {primitive.name}");
                }
                else
                {
                    McpLogger.LogWarning($"[MCP Unity] Parent not found, primitive created at root");
                }
            }

            if (position != null)
            {
                primitive.transform.localPosition = new Vector3(
                    position["x"]?.ToObject<float>() ?? 0f,
                    position["y"]?.ToObject<float>() ?? 0f,
                    position["z"]?.ToObject<float>() ?? 0f
                );
            }

            if (rotation != null)
            {
                primitive.transform.localEulerAngles = new Vector3(
                    rotation["x"]?.ToObject<float>() ?? 0f,
                    rotation["y"]?.ToObject<float>() ?? 0f,
                    rotation["z"]?.ToObject<float>() ?? 0f
                );
            }

            if (scale != null)
            {
                primitive.transform.localScale = new Vector3(
                    scale["x"]?.ToObject<float>() ?? 1f,
                    scale["y"]?.ToObject<float>() ?? 1f,
                    scale["z"]?.ToObject<float>() ?? 1f
                );
            }

            EditorUtility.SetDirty(primitive);

            return new JObject
            {
                ["success"] = true,
                ["type"] = "text",
                ["message"] = $"Created {primitiveType} '{primitive.name}' (instanceId: {primitive.GetInstanceID()})",
                ["data"] = new JObject
                {
                    ["instanceId"] = primitive.GetInstanceID(),
                    ["name"] = primitive.name,
                    ["primitiveType"] = primitiveType.ToString()
                }
            };
        }
    }
}
